addons/survey/controllers/main.py

```diff
@@ -454,6 +458,56 @@ def survey_get_question_image(self, survey_token, answer_token, question_id, sug
             suggested_answer, 'value_image'
         ).get_response()
 
+    @http.route('/survey/<string:survey_token>/preview_picture', type='http', auth='public')
+    def survey_get_preview_picture(self, survey_token, results=False):
+        """ Route used by meta tags for the link preview picture. """
+        survey_sudo, __ = self._fetch_from_access_token(survey_token, False)
+        if not survey_sudo:
+            raise werkzeug.exceptions.NotFound()
+
+        if survey_sudo.certification:
+            tag_description = _("Certification results") if results else _("Certification")
+            tag_name = "trophy"
+        else:
+            tag_description = _("Survey results") if results else _("Survey")
+            tag_name = "survey"
+
+        # PIL does not support vector images formats, fallback to default avatar or no background
+        # image if opening the image failed.
+        if survey_sudo.background_image:
+            background_image_io = BytesIO(b64decode(survey_sudo.background_image))
+            try:
+                background_image = Image.open(background_image_io)
+            except UnidentifiedImageError:
+                background_image = False
+        else:
+            background_image = False
+        if survey_sudo.user_id.image_128:
+            user_image_io = BytesIO(b64decode(survey_sudo.user_id.image_128))
+            try:
+                user_image = Image.open(user_image_io)
+            except UnidentifiedImageError:
+                user_image = Image.open(BytesIO(request.env['avatar.mixin']._avatar_get_placeholder()))
+        else:
+            user_image = Image.open(BytesIO(request.env['avatar.mixin']._avatar_get_placeholder()))
+
+        preview_image = generate_survey_preview_image(
+            background_image=background_image,
+            nb_completed=survey_sudo.answer_done_count,
+            nb_registered=survey_sudo.answer_count,
+            secondary_image=user_image,
+            secondary_text=f"{survey_sudo.user_id.name} - {survey_sudo.create_date.strftime('%b %Y')}",
+            success_rate=survey_sudo.success_ratio,
+            tag_description=tag_description,
+            tag_path=f"survey/static/src/img/{tag_name}-{'light' if background_image else 'brand'}.png",
+            title=survey_sudo.title,
+        )
+
+        output = BytesIO()
+        preview_image.save(output, 'png')
+
+        return request.make_response(output.getvalue(), [('Content-Type', ' image/png')])
```

Comment:
Changed, it now uses odoo/http  `Stream` class which has a `get_response` method, that uses `send_file` 😅

Reply:
Well the docstring of `Stream` is pretty explicit:

> Instantiating a Stream object manually without using one of the
> dedicated constructors is discouraged.

The stream API has 2 purposes: (1) use OS/web server acceleration when streaming *a file stored on the filesystem* to copy the bytes directly from the disk onto the network interface, i.e. without loading anything in RAM, (2) use advanced caching options to let the browser and Odoo know when the resource doesn't need to be transferred again.

Here neither of the two is currently used. The resource is *already* loaded in memory so no acceleration is possible. None of the caching option is set (immutable / max age / etag / last modified) so it will always be transferred.

You can achieve the same result with

```py
return odoo.http.Response(image_apply_opt(preview_image, 'png'), mimetype='image/png')
```

But let's ask the actual questions, how big can that image get? Is it more kb (in which case always sending the image is fine) or it is more Mb (in which case caching may be useful). Could the image be cached on the browser? For how long? Would it be possible to construct a very-fast-to-compute string that would uniquely identify the image (best if we don't actually need to load that image in memory), in such case we could use an ETag.

We can discuss the matter at the office on mercredi.